File: 11. BST 3.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class BuildTree:
    def __init__(self):
        self.idx = 0

    def build_tree(self, arr):
        if self.idx >= len(arr):
            return None

        if  arr[self.idx] == -1:
            self.idx+=1
            return None

        root = Node(arr[self.idx])
        self.idx += 1

        root.left = self.build_tree(arr)
        root.right = self.build_tree(arr)

        return root

    def preorder(self, root):
        if root is None:
            return
        print(root.data, end=" ")
        self.preorder(root.left)
        self.preorder(root.right)

    def inorder(self, root):
        if root is None:
            return
        self.inorder(root.left)
        print(root.data, end=" ")
        self.inorder(root.right)

    def postorder(self, root):
        if root is None:
            return
        self.postorder(root.left)
        self.postorder(root.right)
        print(root.data, end=" ")


import sys
arr = list(map(int, sys.stdin.read().split()))
bt = BuildTree()
root = bt.build_tree(arr)
bt.preorder(root)
print()
bt.inorder(root)
print()
bt.postorder(root)

File: 11. BST 3.java
import java.util.*;

class Node 
{
    int data;
    Node left;
    Node right;
    Node(int data) 
    {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}


class BuildTree 
{
    int idx = 0;
    public Node buildTree(List<Integer> arr) 
    {
        if (idx >= arr.size() || arr.get(idx) == -1) 
        {
            idx++;
            return null;
        }

        Node root = new Node(arr.get(idx++));
        root.left = buildTree(arr);
        root.right = buildTree(arr);

        return root;
    }
    public void preorder(Node root) 
    {
        if (root == null) return;
        System.out.print(root.data + " ");
        preorder(root.left);
        preorder(root.right);
    }

    public void inorder(Node root) 
    {
        if (root == null) return;
        inorder(root.left);
        System.out.print(root.data + " ");
        inorder(root.right);
    }

    public void postorder(Node root) 
    {
        if (root == null) return;
        postorder(root.left);
        postorder(root.right);
        System.out.print(root.data + " ");
    }
}

public class Main 
{
    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        List<Integer> arr = new ArrayList<>();
        while (sc.hasNextInt()) 
        {
            arr.add(sc.nextInt());
        }
        BuildTree bt = new BuildTree();
        Node root = bt.buildTree(arr);

        bt.preorder(root);
        System.out.println();

        bt.inorder(root);
        System.out.println();

        bt.postorder(root);
    }
}
